### DISCUSSION TITLE
YCbCr (I420) -> RGB conversion with libvips is a significant bottleneck

### DISCUSSION THREAD

**Discussion by midn1** [libvips/libvips#2561] [Q&A] [answered]
Sorry to bother, and this isn't really urgent (at least not for a while).

I am retrieving YCbCr images that I need to load into VIPS for further processing. Without the YCC->RGB conversion I get ~0.0016s/frame, while with - 0.17s/frame, which is a little over 100 times slower. Ideally I'd like to have it over 30FPS to leave time for the actual processing.

Here's the relevant code:
```c
VipsImage *yuvsep[3] = {
	vips_image_new_from_memory(img->planes[VPX_PLANE_Y], img->stride[VPX_PLANE_Y] * img->d_h, img->stride[VPX_PLANE_Y], img->d_h, 1, VIPS_FORMAT_UCHAR),
	vips_image_new_from_memory(img->planes[VPX_PLANE_U], img->stride[VPX_PLANE_U] * (img->d_h >> img->y_chroma_shift), img->stride[VPX_PLANE_U], img->d_h >> img->y_chroma_shift, 1, VIPS_FORMAT_UCHAR),
	vips_image_new_from_memory(img->planes[VPX_PLANE_V], img->stride[VPX_PLANE_V] * (img->d_h >> img->y_chroma_shift), img->stride[VPX_PLANE_V], img->d_h >> img->y_chroma_shift, 1, VIPS_FORMAT_UCHAR)
};

vips_crop(yuvsep[0], &yuvsep[0], 0, 0, img->d_w, img->d_h, NULL);
vips_linear1(yuvsep[0], &yuvsep[0], 1, -16, NULL);

vips_crop(yuvsep[1], &yuvsep[1], 0, 0, img->d_w >> img->x_chroma_shift, img->d_h >> img->y_chroma_shift, NULL);
vips_resize(yuvsep[1], &yuvsep[1], 1 << img->x_chroma_shift, "vscale", (double) (1 << img->y_chroma_shift), NULL);
vips_linear1(yuvsep[1], &yuvsep[1], 1, -128, NULL);

vips_crop(yuvsep[2], &yuvsep[2], 0, 0, img->d_w >> img->x_chroma_shift, img->d_h >> img->y_chroma_shift, NULL);
vips_resize(yuvsep[2], &yuvsep[2], 1 << img->x_chroma_shift, "vscale", (double) (1 << img->y_chroma_shift), NULL);
vips_linear1(yuvsep[2], &yuvsep[2], 1, -128, NULL);

VipsImage *yuv;
vips_bandjoin(yuvsep, &yuv, 3, NULL);
vips_copy(yuv, &yuv, "interpretation", VIPS_INTERPRETATION_MULTIBAND, NULL); /* Necessary, otherwise it's considered black & white. */

VipsImage *rgb;
vips_recomb(yuv, &rgb, YCC2RGB, NULL);
```

Please excuse my lack of memory handling and gamma correction - I'm doing things one step at a time.

Due to the fact that the planes have padding between rows (which libvips doesn't seem to support), I have to load them with the padding before cropping. The images are in 4:2:0, so the Cb and Cr planes need to then be scaled up. Then the actual format conversion takes place.

According to my tests, the recombination takes up almost 25% of the time. So I'm not sure how to continue. I can try rewriting with integer arithmetic (which VIPS doesn't make very easy, as `vips_linear(1)` gives either float or uchar), but I'm also not too certain about how the quality would be.

This is the conversion formula I'm working with, although the matrix coefficients may differ:
![image](https://wikimedia.org/api/rest_v1/media/math/render/svg/16b6c7a2fed82e2b46729873c51268ed32c10832)

Thanks in advance.

**Comment by jcupitt** [comment, marked as answer]
Hi @midn1,

That's an interesting problem. I tried to do what you're doing in python for easy testing:

```python
#!/usr/bin/python3

import sys
import time
import pyvips

if len(sys.argv) != 3:
    print("usage: {sys.argv[0]) WIDTH HEIGHT")
    sys.exit(1)

width = int(sys.argv[1])
height = int(sys.argv[2])
shift = 1
subsample = 1 << shift
ycc2rgb = [
    [1.0,  0.0,       1.402   ],
    [1.0, -0.344136, -0.714136],
    [1.0,  1.772,     0.0     ],
]

Y = pyvips.Image.black(width, height).copy_memory()
Cb = pyvips.Image.black(width >> shift, height >> shift).copy_memory()
Cr = pyvips.Image.black(width >> shift, height >> shift).copy_memory()

# original
def try1(Y, Cb, Cr):
    Y -= 16
    Cb = Cb.resize(subsample) - 128
    Cr = Cr.resize(subsample) - 128
    YCbCr = Y.bandjoin([Cb, Cr])
    return YCbCr.recomb(ycc2rgb).copy(interpretation="srgb")

# use linear inerp, plus multi-valued linear 
def try2(Y, Cb, Cr):
    Cb = Cb.resize(subsample, kernel="linear")
    Cr = Cr.resize(subsample, kernel="linear")
    YCbCr = Y.bandjoin([Cb, Cr]) - [16, 128, 128]
    return YCbCr.recomb(ycc2rgb).copy(interpretation="srgb")

# try without recomb
def try3(Y, Cb, Cr):
    Cb = Cb.resize(subsample, kernel="linear")
    Cr = Cr.resize(subsample, kernel="linear")
    YCbCr = Y.bandjoin([Cb, Cr]) - [16, 128, 128]
    return YCbCr.copy(interpretation="srgb")

def bench(msg, fn, Y, Cb, Cr):
    start = time.time()
    result = fn(Y, Cb, Cr).copy_memory()
    end = time.time()
    print(f"{msg}: {1000 * (end - start):.1f}ms")

bench("original", try1, Y, Cb, Cr)
bench("linear", try2, Y, Cb, Cr)
bench("no recomb", try3, Y, Cb, Cr)
```

I see:

```
$ ./try334.py 5000 5000
original: 272.5ms
linear: 166.4ms
no recomb: 115.4ms
```

So for me the recomb is taking about 20% of the time, and the cubic upscale (the default for resize) about 50%. 

With a video frame I see:

```
$ ./try334.py 640 480
original: 15.1ms
linear: 6.9ms
no recomb: 6.6ms
```

Which should be fast enough for 30fps, though of course this is a fast PC.

To go quicker, I think you'd need to write a custom libvips operator to do your subsampled YCbCr -> RGB conversion. Did you see the section in the manual on this? It should be simple, and you can borrow optimisation ideas from code like libjpeg-turbo, which has a very highly tuned ycbcr -> rgb conversion.

https://www.libvips.org/API/current/extending.html

**Reply by midn1** [reply]
Gah, I feel stupid. Indeed, it works great now. Thank you for your help and work.

**Comment by jcupitt** [comment]
Ooop, I was using a debug build of libvips by mistake. If I enable `-O3` I see:

```
$ ./try334.py 5000 5000
original: 113.6ms
linear: 83.1ms
no recomb: 71.7ms
```

About 2x faster. It might be worth checking how your libvips was built.
